src/cli.py

```diff
@@ -42,479 +104,1158 @@ def _format_action_invocation(self, action):
-
-        # deal with options set in user-specified defaults file, if present
-        config_path = self.config.get('INPUT_FILE', None)
-        config_str = ''
-        if config_path:
+                yield from arg_gp.arguments
+
+    def add_input_file_arg(self, target_p):
+        """Convenience method to add the flag to pass a user-designated defaults
+        file to the parser ``target_p`` (either the top-level parser, or the 
+        preparser.)
+        """
+        kwargs = {'type': str}
+        if isinstance(target_p, MDTFTopLevelArgParser):
+            kwargs.update({
+                'metavar': "INPUT_FILE",
+                'help': word_wrap("""
+                    Path to a user configuration file that sets options listed 
+                    here. This can be a JSONC file of the form given in 
+                    sample_input.jsonc, or a text file containing command-line 
+                    arguments. Options set explicitly on the command line will 
+                    still override settings in this file.
+                """)
+            })     
+        target_p.add_argument('--input_file', '--input-file', '-f', **kwargs)
+
+    def init_user_defaults(self):
+        """Set user defaults using values read in from a configuration
+        file in one of two formats. 
+        
+        Args:
+            config_str (str): contents of the configuration file, either:
+
+            1. A JSON/JSONC file of key-value pairs. This is parsed using
+                :func:`~src.util.filesystem.parse_json`.
+            2. A plain text file containing flags and arguments as they would
+                be passed on the command line (except shell expansions are not 
+                performed). This is parsed by the :meth:`MDTFArgParser.parse_args`
+                method of the configured parser.
+
+        The format is determined from context. ValueError is raised if the string
+        cannot be parsed.
+        """
+        config = CLIConfigManager()
+        input_p = MDTFArgPreparser()
+        self.add_input_file_arg(input_p)
+        path = input_p.parse_input_file(self.argv)
+        if not path:
+            return
+        try:
+            with io.open(path, 'r', encoding='utf-8') as f:
+                str_ = f.read()
+        except Exception:
+            _log.exception("Can't read user input file at %s.", path)
+            raise ValueError()
+        if not str_:
+            return
+        # try to determine if file is json
+        if 'json' in os.path.splitext(path)[1].lower():
+            # assume config_file a JSON dict of option:value pairs.
             try:
-                with io.open(config_path, 'r', encoding='utf-8') as f:
-                    config_str = f.read()
+                d = util.parse_json(str_)
+                self.file_case_list = d.pop('case_list', [])
+                # 'pods' handled by normal CLI
+                # self.file_pod_list = d.pop('pods', [])
```

Comment:
Delete commented-out line?

Reply:
Done in NOAA-GFDL/MDTF-diagnostics@7b47104f5077872e6b20e1af63fb487da281db9b